File: src/engine/validation.ts
/**
 * Authoring-time validation (v1.4 §15.5, §3.3; Rebuild_Brief §5).
 * Rejects subscriptions to non-canonical events, keyless locks, empty shield
 * rows, and malformed configs — config errors, not gameplay states.
 */
import type { CardDefinition, EncounterConfig, InfoNugget, TriggerCondition } from './types';
import { CANONICAL_EVENTS } from './types';

export interface ValidationIssue {
  severity: 'error' | 'warning';
  where: string;
  message: string;
}

const isCanonical = (e: string): boolean => (CANONICAL_EVENTS as readonly string[]).includes(e);

function validateTrigger(t: TriggerCondition, where: string, issues: ValidationIssue[]): void {
  if (!isCanonical(t.event)) {
    issues.push({
      severity: 'error',
      where,
      message: `Trigger subscribes to non-canonical event "${t.event}". Canonical events: ${CANONICAL_EVENTS.join(', ')}.`,
    });
  }
}

export function validateCard(card: CardDefinition): ValidationIssue[] {
  const issues: ValidationIssue[] = [];
  const w = `card ${card.id}`;

  if (card.subtype === 'Trap') {
    if (!card.trapTrigger) {
      issues.push({ severity: 'error', where: w, message: 'Trap card has no trapTrigger — no silent dead traps (v1.4 §3.6).' });
    } else {
      validateTrigger(card.trapTrigger, w, issues);
    }
    if (!card.keywords.includes('Trap')) {
      issues.push({ severity: 'warning', where: w, message: 'Trap subtype without Trap keyword.' });
    }
  }
  for (const ab of card.triggeredAbilities ?? []) {
    validateTrigger(ab.trigger, `${w} / ability ${ab.id}`, issues);
  }
  if (card.keywords.includes('Rapport') && !card.rapport) {
    issues.push({ severity: 'error', where: w, message: 'Rapport keyword requires rapport prediction config (v1.4 §8.3).' });
  }
  if (card.keywords.includes('Heavy Hand') && !card.heavyHandEffects) {
    issues.push({ severity: 'error', where: w, message: 'Heavy Hand keyword requires heavyHandEffects (v1.4 §8.3).' });
  }
  if (card.supertype === 'Information' && !card.nuggetId) {
    issues.push({ severity: 'error', where: w, message: 'Information Card must carry a nuggetId (v1.4 §3.9).' });
  }
  if (card.subtype === 'Token' && card.cost !== 0) {
    issues.push({ severity: 'warning', where: w, message: 'Tokens are never played from hand; cost is meaningless.' });
  }
  return issues;
}

export function validateEncounter(
  config: EncounterConfig,
  cards: Record<string, CardDefinition>,
  nuggets: Record<string, InfoNugget>,
): ValidationIssue[] {
  const issues: ValidationIssue[] = [];
  const w = `encounter ${config.id}`;

  // ≥ 1 opponent shield total (v1.4 §3.3 validation; Brief §7 trap 12 — no vacuous wins)
  if (config.npcGuardShieldCount + config.opponentShields.length < 1) {
    issues.push({ severity: 'error', where: w, message: 'Encounter must define at least one opponent shield (guards + cores ≥ 1).' });
  }
  if (config.npcGuardShieldCount < 0) {
    issues.push({ severity: 'error', where: w, message: 'npcGuardShieldCount cannot be negative.' });
  }

  for (const [i, shield] of config.opponentShields.entries()) {
    const sw = `${w} / core shield ${i} (${shield.cardId})`;
    if (!shield.keyNuggetIds || shield.keyNuggetIds.length === 0) {
      issues.push({ severity: 'error', where: sw, message: 'NPC Core Shield lists no key nuggets — a keyless lock is a config error (v1.4 §3.3).' });
    }
    for (const nid of shield.keyNuggetIds ?? []) {
      if (!nuggets[nid]) {
        issues.push({ severity: 'error', where: sw, message: `Key nugget "${nid}" does not exist.` });
      }
    }
    if (!shield.isHint && !cards[shield.cardId]) {
      issues.push({ severity: 'error', where: sw, message: `Core shield card "${shield.cardId}" not found (non-Hint shields add their card to the Collection).` });
    }
  }

  for (const id of config.enemyDeckCardIds) {
    if (!cards[id]) issues.push({ severity: 'error', where: w, message: `Enemy deck card "${id}" not found.` });
  }
  for (const sp of config.scheduledPlays ?? []) {
    if (!config.enemyDeckCardIds.includes(sp.cardId)) {
      issues.push({ severity: 'error', where: w, message: `Scheduled play "${sp.cardId}" is not in the enemy deck.` });
    }
  }
  for (const id of config.startingImpressions ?? []) {
    if (!cards[id]) issues.push({ severity: 'error', where: w, message: `Starting impression "${id}" not found.` });
  }
  for (const ov of config.nuggetOverrides) {
    if (!nuggets[ov.nuggetId]) {
      issues.push({ severity: 'error', where: w, message: `Nugget override references unknown nugget "${ov.nuggetId}".` });
    }
  }
  if (config.playerDummyShieldSlots < 0) {
    issues.push({ severity: 'error', where: w, message: 'playerDummyShieldSlots cannot be negative.' });
  }
  return issues;
}

export function assertValid(issues: ValidationIssue[]): void {
  const errors = issues.filter((i) => i.severity === 'error');
  if (errors.length > 0) {
    throw new Error(`Validation failed:\n${errors.map((e) => `  [${e.where}] ${e.message}`).join('\n')}`);
  }
}

File: csharp-engine/Breakthrough.Engine/Validation.cs
// Authoring-time validation (v1.4 §15.5, §3.3; Rebuild_Brief §5).
// Rejects subscriptions to non-canonical events, keyless locks, empty shield
// rows, and malformed configs — config errors, not gameplay states.
//
// 1:1 C# port of src/engine/validation.ts (messages preserved verbatim).

namespace Breakthrough.Engine;

public sealed record ValidationIssue(string Severity, string Where, string Message);

public static class Severities
{
    public const string Error = "error";
    public const string Warning = "warning";
}

public sealed class ValidationException(string message) : Exception(message);

public static class Validation
{
    private static bool IsCanonical(string e) => EventTypes.Canonical.Contains(e);

    private static void ValidateTrigger(TriggerCondition t, string where, List<ValidationIssue> issues)
    {
        if (!IsCanonical(t.Event))
        {
            issues.Add(new ValidationIssue(
                Severities.Error,
                where,
                $"Trigger subscribes to non-canonical event \"{t.Event}\". Canonical events: {string.Join(", ", EventTypes.Canonical)}."));
        }
    }

    public static List<ValidationIssue> ValidateCard(CardDefinition card)
    {
        var issues = new List<ValidationIssue>();
        string w = $"card {card.Id}";

        if (card.Subtype == Subtypes.Trap)
        {
            if (card.TrapTrigger == null)
            {
                issues.Add(new ValidationIssue(Severities.Error, w, "Trap card has no trapTrigger — no silent dead traps (v1.4 §3.6)."));
            }
            else
            {
                ValidateTrigger(card.TrapTrigger, w, issues);
            }
            if (!card.Keywords.Contains(Keywords.Trap))
            {
                issues.Add(new ValidationIssue(Severities.Warning, w, "Trap subtype without Trap keyword."));
            }
        }
        foreach (var ab in card.TriggeredAbilities ?? [])
        {
            ValidateTrigger(ab.Trigger, $"{w} / ability {ab.Id}", issues);
        }
        if (card.Keywords.Contains(Keywords.Rapport) && card.Rapport == null)
        {
            issues.Add(new ValidationIssue(Severities.Error, w, "Rapport keyword requires rapport prediction config (v1.4 §8.3)."));
        }
        if (card.Keywords.Contains(Keywords.HeavyHand) && card.HeavyHandEffects == null)
        {
            issues.Add(new ValidationIssue(Severities.Error, w, "Heavy Hand keyword requires heavyHandEffects (v1.4 §8.3)."));
        }
        if (card.Supertype == Supertypes.Information && card.NuggetId == null)
        {
            issues.Add(new ValidationIssue(Severities.Error, w, "Information Card must carry a nuggetId (v1.4 §3.9)."));
        }
        if (card.Subtype == Subtypes.Token && card.Cost != 0)
        {
            issues.Add(new ValidationIssue(Severities.Warning, w, "Tokens are never played from hand; cost is meaningless."));
        }
        return issues;
    }

    public static List<ValidationIssue> ValidateEncounter(
        EncounterConfig config,
        IReadOnlyDictionary<string, CardDefinition> cards,
        IReadOnlyDictionary<string, InfoNugget> nuggets)
    {
        var issues = new List<ValidationIssue>();
        string w = $"encounter {config.Id}";

        int guardCount = EncounterDefaults.ResolvedGuardCount(config);
        // ≥ 1 opponent shield total (v1.4 §3.3 validation; Brief §7 trap 12 — no vacuous wins)
        if (guardCount + config.OpponentShields.Count < 1)
        {
            issues.Add(new ValidationIssue(Severities.Error, w, "Encounter must define at least one opponent shield (guards + cores ≥ 1)."));
        }
        if (guardCount < 0)
        {
            issues.Add(new ValidationIssue(Severities.Error, w, "npcGuardShieldCount cannot be negative."));
        }

        // v1.4.1 — card-backed Guard Shields count toward the guard total.
        var guardCards = config.NpcGuardShieldCardIds ?? [];
        if (guardCards.Count > guardCount)
        {
            issues.Add(new ValidationIssue(
                Severities.Error,
                w,
                $"npcGuardShieldCardIds ({guardCards.Count}) exceeds the guard total ({guardCount}) — card guards count toward the total."));
        }
        foreach (var id in guardCards)
        {
            if (!cards.ContainsKey(id))
            {
                issues.Add(new ValidationIssue(Severities.Error, w, $"Guard Shield card \"{id}\" not found."));
            }
            else if (!cards[id].Keywords.Contains(Keywords.ShieldTrigger))
            {
                issues.Add(new ValidationIssue(Severities.Warning, w, $"Guard Shield card \"{id}\" has no Shield Trigger — it will break with no effect."));
            }
        }

        for (int i = 0; i < config.OpponentShields.Count; i++)
        {
            var shield = config.OpponentShields[i];
            string sw = $"{w} / core shield {i} ({shield.CardId})";
            if (shield.KeyNuggetIds == null || shield.KeyNuggetIds.Count == 0)
            {
                issues.Add(new ValidationIssue(Severities.Error, sw, "NPC Core Shield lists no key nuggets — a keyless lock is a config error (v1.4 §3.3)."));
            }
            foreach (var nid in shield.KeyNuggetIds ?? [])
            {
                if (!nuggets.ContainsKey(nid))
                {
                    issues.Add(new ValidationIssue(Severities.Error, sw, $"Key nugget \"{nid}\" does not exist."));
                }
            }
            if (!shield.IsHint && !cards.ContainsKey(shield.CardId))
            {
                issues.Add(new ValidationIssue(Severities.Error, sw, $"Core shield card \"{shield.CardId}\" not found (non-Hint shields add their card to the Collection)."));
            }
        }

        foreach (var id in config.EnemyDeckCardIds)
        {
            if (!cards.ContainsKey(id)) issues.Add(new ValidationIssue(Severities.Error, w, $"Enemy deck card \"{id}\" not found."));
        }
        foreach (var sp in config.ScheduledPlays ?? [])
        {
            if (!config.EnemyDeckCardIds.Contains(sp.CardId))
            {
                issues.Add(new ValidationIssue(Severities.Error, w, $"Scheduled play \"{sp.CardId}\" is not in the enemy deck."));
            }
        }
        foreach (var id in config.StartingImpressions ?? [])
        {
            if (!cards.ContainsKey(id)) issues.Add(new ValidationIssue(Severities.Error, w, $"Starting impression \"{id}\" not found."));
        }
        foreach (var ov in config.NuggetOverrides)
        {
            if (!nuggets.ContainsKey(ov.NuggetId))
            {
                issues.Add(new ValidationIssue(Severities.Error, w, $"Nugget override references unknown nugget \"{ov.NuggetId}\"."));
            }
        }
        if (EncounterDefaults.ResolvedDummySlots(config) < 0)
        {
            issues.Add(new ValidationIssue(Severities.Error, w, "playerDummyShieldSlots cannot be negative."));
        }
        return issues;
    }

    public static void AssertValid(IEnumerable<ValidationIssue> issues)
    {
        var errors = issues.Where(i => i.Severity == Severities.Error).ToList();
        if (errors.Count > 0)
        {
            throw new ValidationException(
                $"Validation failed:\n{string.Join("\n", errors.Select(e => $"  [{e.Where}] {e.Message}"))}");
        }
    }
}
